doc/adr/0007-recipe-secrets.md

```diff
@@ -0,0 +1,49 @@
+# Handing of Recipe Secrets
+
+## Status
+Proposed
+
+## Context
+
+Recipes often need to pull data from download locations (e.g. FTP servers) that require authentication with secres (e.g. username / password).
+We need a robust and secure way to pass these secrets from the recipe maintainer to the bakery.
+
+## Decision
+
+### Storing Secrets in Feedstocks
+
+We will add an additional optional configuration file to the recipe feedstock called `secrets.yaml.gpg`.
+The unencrypted version of the file might look something like this
+```yaml
+username: Foo
+password: Bar
+```
+
+Pangeo Forge will create a GPG key pair and publish the public key under the idenity `secrets@pangeo-forge.org`.
+Users will [encrypt](https://www.gnupg.org/gph/en/manual/x110.html) the secrets file using gpg on the command line as follows
+```bash
+gpg --output secrets.yaml.gpg --encrypt --recipient secrets@pangeo-forge.org secrets.yaml
+```
+and add this file to their feedstock
+
+The CI environment will have access to the private key and will be able to decrypt the file and read the secrets.
+`pangeo_forge_prefect` will inject any secrets found in the file into the flow runtime using the [Prefect Client Secrets API](https://docs.prefect.io/orchestration/concepts/secrets.html#prefect-client).
+
+When defining the recipe, users will no longer specify plain-text passwords, instead they will use a new `pangeo_forge_recipes.Secret` type.
+
+```python
+pattern = FilePattern(
+    ...,
+    fsspec_open_kwargs={"auth": aiohttp.BasicAuth(Secret['username'], Secret['password'])},
+)
+```
+
+The `Secret` type will subclass `str` but will load its values from either environment variables or Prefect Secrets.
```

Comment:
Here is a potential implementation

```python
import os
import pickle

class Secret(str):
    
    @classmethod
    def _get_secret_from_env(cls, env_var_name):
        try:
            from prefect.client import Secret as PrefectSecret
            return PrefectSecret(env_var_name).get()
        except (ImportError, ValueError):
            return os.environ[env_var_name]
    
    def __new__(cls, env_var_name):
        val = cls._get_secret_from_env(env_var_name)
        return super().__new__(cls, val) 
    
    def __init__(self, env_var_name):
        self._env_var_name = env_var_name
    
    def __reduce__(self):
        return (self.__class__, (self._env_var_name, ))
    
os.environ["FOO"] = "BAR"
s = Secret("FOO")
assert s == "BAR"
assert s._env_var_name == "FOO"
p = pickle.dumps(s)
os.environ["FOO"] = "BAZ"
s_ = pickle.loads(p)
assert s == "BAR"
assert s_ == "BAZ"
```

Reply:
Would this go in `pangeo_forge_recipes.storage`?